lib/mix/lib/mix/tasks/run.ex

```diff
@@ -111,10 +111,11 @@ defmodule Mix.Tasks.Run do
       end)
 
     {file, argv} =
-      case {Keyword.has_key?(opts, :eval), head} do
-        {true, _} -> {nil, head}
-        {_, [head | tail]} -> {head, tail}
-        {_, []} -> {nil, []}
+      case {Keyword.has_key?(opts, :eval), args, head} do
+        {true, _, _} -> {nil, head}
+        {_, ["--" | _], head} -> {nil, head}
```

Comment:
Gotcha so maybe checking for `opts` makes more sense.

Reply:
Maybe the issue is with option parser swallowing `—`?

Reply:
`OptionParser.parse_head!` is returning:

```
iex> OptionParser.parse_head!(["--", "foo"], strict: [halt: :boolean])                        
{[], ["foo"]}

iex> OptionParser.parse_head!(["--no-halt", "--", "foo"], strict: [halt: :boolean])           
{[halt: false], ["foo"]}

iex> OptionParser.parse_head!(["--no-halt", "foo.exs", "--", "foo"], strict: [halt: :boolean])
{[halt: false], ["foo.exs", "--", "foo"]}
```

Do you mean option parser should return `~w(-- foo)` for the first 2 cases? Even if that's the case the current `def run` code will need a fix because it's swapping the first item as the file so `~w(-- foo)` would resolve the file as `--`. I pushed https://github.com/elixir-lang/elixir/pull/12499/commits/8f7772bbe8422399569f9f7c126c1b81ca7d5ae7 testing more cases.

Btw that's the behavior when calling elixir:

```
# Elixir 1.14.3
❯ elixir --no-halt foo.exs -- foo
["--", "foo"]
```

Reply:
The tests look great, so we need to make them work based exclusively on the result of OptIonParser (and not its original args). I don’t think that possible though, so we will need to add an option so OptIonParser returns the double dash if it finds it.

Reply:
We can call it :return_separator and it defaults to false. :)

Reply:
Sounds good, I'm gonna work on it thanks